dashdotdb/controllers/tokens.py

```diff
@@ -0,0 +1,44 @@
+import datetime
+
+from uuid import uuid4
+
+from dashdotdb.models.dashdotdb import db
+from dashdotdb.models.dashdotdb import Tokens
+from dashdotdb.models.dashdotdb import LatestTokens
+
+
+def close_token(uuid, data_type):
+    db_token = db.session.query(Tokens) \
+        .filter(Tokens.uuid == uuid, Tokens.data_type == data_type).first()
+    if db_token is None:
+        return 'token not found', 404
+    db.session.update(Tokens).where(
+        Tokens.id == db_token.id).values(is_open=False)
+    latest_token = db.session.query(LatestTokens) \
+        .filter(Tokens.data_type == data_type).first()
+    if latest_token is None:
+        db.session.add(LatestTokens(data_type=data_type, token_id=db_token.id))
+    else:
+        # Only update the latest token if the creation timestamp is newer than
+        # the current latest token creation timestamp
+        latest_token_data = db.session.query(
+            Tokens).filter(Tokens.id == latest_token.id)
+        if db_token.creation_timestamp > latest_token_data.createion_timestamp:
+            db.session.update(LatestTokens).where(
+                data_type=data_type).values(token_id=db_token.id)
+    db.session.commit()
```

Comment:
This should be inside a transaction, to avoid race conditions

Reply:
If I understand sqlalchemy correctly, this is being done in a transaction.  All actions against a session are in a transaction and only hit the DB when commit() is called:

https://docs.sqlalchemy.org/en/14/orm/session_transaction.html

Reply:
while I'm sure that a virtual transaction is created whenever you do an update as autocommit is disabled, I doubt very much that sqlalchemy is going to create a transaction before a query unless you explicitly state it.

Reply:
Explicitly started a transaction